requests-for-comments/2024-01-23-project-environments.md

```diff
@@ -0,0 +1,116 @@
+
+When doing that, customers typically want to share setup between environments, while ensuring *data* from one environment doesn't pollute the others.
+
+The multiple environments use case is one of the reasons why projects are a feature of PostHog, and projects are great for tracking separate products, but not ideal for tracking multiple instances of the same product: whatever you set up for one project, you have to manually replicate in all others.
+
+We currently solve this in Feature flags with a feature for syncing flags between projects, based on the flag key. Feature flags make this uniquely easy thanks to the feature flag key being an effective organization-wide identifier. No other entity – dashboards, insights, etc. – has this advantage though.
+
+Because feature flags use cohorts and actions in their definitions, these are also synced in the current solution –but only based on the cohort/action name, so if the name is changed in one environment, the link is silently broken.
+
+
+Previous context on this topic: [the December 2022 "Feature flag environments" RFC](https://github.com/PostHog/product-internal/blob/main/requests-for-comments/2022-12-08-feature-flag-environments.md#221115---rfc-feature-flag-environments)
+
+## Expectations
+
+### Isolated
+
+The key feature is separation of all of *data*:
+
+- Events
+- People
+- Groups
+- Recordings
+
+Pipeline setup should probably be separate too:
+
+- Apps
+- Batch exports
+
+Domain-specific configuration is also going to differ between environments likely:
+
+- Toolbar authorized URLs
+
+### Shared
+
+Sharing of these entities is commonly requested:
+
+- Dashboards
+- Insights
+- Actions
+- Cohorts
+- Feature flags
+
+Not mentioned typically, but logically also shared should be:
+
+- Event and property definitions
+- Group types
+- Annotations
+- Playlists
+- Experiments
+- Surveys
+- Early access features
+- Project settings
+
+The needs are unclear for:
+
+- Data warehouse tables
+
+## Rejected solution: Cross-project copying
+
+This is what some users have asked for verbatim. What it would probably mean is a feature flags-style "Projects" view on the item level, allowing an item to be copied/synced across projects. The reliability of syncing is key though, and the problem mentioned before is that there's no inherent cross-project ID for this. Just basing on the entity name would be extremely fragile – there needs to be an underlying link better than that.
+
+This means we'd need to maintain such underlying links (e.g. a cross-project ID, adjacent to the primary key ID) for all entities we'd like to work cross-project. In one way, this could be nice if you'd like to develop, say, a dashboard in the test environment and only copy it to the production environment when final. Divergence becomes easy. But there would be an act of manual syncing, making the experience cumbersome for users at scale. This would work, but it wouldn't be turnkey.
```

Comment:
Not saying we shouldn't reject this solution, but one important thing, at least for flags is that divergence is a feature not a bug - you would want to freely edit a flag once you've copied it to a dev environment to test out flows / do random things - and you don't want any of these to reflect in the prod environment, for whatever reason.

So the usual flow I've seen so far is:
1. Setup flag as we'd like in prod
2. Sync to dev
3. Test in dev, do debugging in dev (editing flag)

Reply:
Although, would be pretty nice if there was a way to reset the dev flag to be back to the prod-like state

Reply:
Do you think we could make rollout groups optionally environment-specific, assuming that the API token is uniquely assigned to an environment? This would be trivial with decide as far as I can tell, but I know less about what info local evaluation has

Reply:
Yep local evaluation does the same, it's passed only one set of release conditions based on the token, and _we could_ do the same as decide here to evaluate the right version of flags.

Reply:
although, want to dig deeper before we go down this route, because right now this feels like a fancier way of replicating the copy to project button 😅

Reply:
For feature flags, it very much is a fancy way of replicating the "copy to project" button. Users experience this problem on a scale much broader than just feature flags though, hence the RFC. This is one of these problems that needs a product-wide solution – if we were to solve it separately for feature flags _and_ dashboards/insights _and_ cohorts/actions _and_ taxonomy, we'd end up with grand patchwork.

Thanks for pointing out divergence as a feature! In the rewritten RFC I propose we make rollout groups environment-specific